### PR TITLE
Add IJEPA task

### PR DESCRIPTION
**Pull request by vahid0001**
# PR Type
[Feature]

# Short Description
Add first version of IJEPA pretraining task

# Tests Added
N/A

- - -
This change is [Reviewable](https://reviewable.io/reviews/VectorInstitute/mmlearn/25)

### CODE REVIEW

**Review by fcogidi — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/VectorInstitute/mmlearn/25)* status: 0 of 1 files reviewed, 14 unresolved discussions (waiting on @vahid0001)

___
*[`mmlearn/tasks/ijepa_pretraining.py` line 19 at r1](https://reviewable.io/reviews/VectorInstitute/mmlearn/25#-O8wXtlU1M_ZA22MVLiV:-O8wXtlU1M_ZA22MVLiW:undefined) ([raw file](https://github.com/VectorInstitute/mmlearn/blob/be3ac266a50f4884bc2d5cbf471cd43aab1a2cef/mmlearn/tasks/ijepa_pretraining.py#L19)):*
> ```Python
> 
> 
> class IJEPAPretraining(L.LightningModule):
> ```





_Suggestion:_
```Python
class IJEPA(L.LightningModule):
```
___
*[`mmlearn/tasks/ijepa_pretraining.py` line 38 at r1](https://reviewable.io/reviews/VectorInstitute/mmlearn/25#-O8wa8lI62Z80MQk29tG:-O8wa8lI62Z80MQk29tH:undefined) ([raw file](https://github.com/VectorInstitute/mmlearn/blob/be3ac266a50f4884bc2d5cbf471cd43aab1a2cef/mmlearn/tasks/ijepa_pretraining.py#L38)):*
> ```Python
>     pred_depth : int
>         Depth of the predictor.
>     optimizer : Optional[Any], optional
> ```





_Suggestion:_
```Python
optimizer : Optional[torch.optim.Optimizer], optional
```
___
*[`mmlearn/tasks/ijepa_pretraining.py` line 39 at r1](https://reviewable.io/reviews/VectorInstitute/mmlearn/25#-O8wa1xLBGzSXEldSens:-O8wa1xMANC-P_fHXv-d:b-28y2a7) ([raw file](https://github.com/VectorInstitute/mmlearn/blob/be3ac266a50f4884bc2d5cbf471cd43aab1a2cef/mmlearn/tasks/ijepa_pretraining.py#L39)):*
> ```Python
>         Depth of the predictor.
>     optimizer : Optional[Any], optional
>         Optimizer configuration, by default None.
> ```

What you will get here is not a config, but an initialized optimizer.
___
*[`mmlearn/tasks/ijepa_pretraining.py` line 41 at r1](https://reviewable.io/reviews/VectorInstitute/mmlearn/25#-O8waHbm4dewKpLxdUHQ:-O8waHbm4dewKpLxdUHR:b9t01ra) ([raw file](https://github.com/VectorInstitute/mmlearn/blob/be3ac266a50f4884bc2d5cbf471cd43aab1a2cef/mmlearn/tasks/ijepa_pretraining.py#L41)):*
> ```Python
>         Optimizer configuration, by default None.
>     lr_scheduler : Optional[Any], optional
>         Learning rate scheduler configuration, by default None.
> ```

Same as the optimizer. This will be an instantiated learning rate scheduler, if one is provided.
___
*[`mmlearn/tasks/ijepa_pretraining.py` line 91 at r1](https://reviewable.io/reviews/VectorInstitute/mmlearn/25#-O8waS7j8IWWRVmfzL95:-O8waS7j8IWWRVmfzL96:bvlmyx6) ([raw file](https://github.com/VectorInstitute/mmlearn/blob/be3ac266a50f4884bc2d5cbf471cd43aab1a2cef/mmlearn/tasks/ijepa_pretraining.py#L91)):*
> ```Python
>         self.total_steps = None
> 
>         self.encoder = VisionTransformer.__dict__[model_name](
> ```

Like we discussed. I think this should be passed in already initialized. Same with the predictor.
___
*[`mmlearn/tasks/ijepa_pretraining.py` line 106 at r1](https://reviewable.io/reviews/VectorInstitute/mmlearn/25#-O8wabwuEmfD5pIsFYtB:-O8wabwuEmfD5pIsFYtC:b-pwoawx) ([raw file](https://github.com/VectorInstitute/mmlearn/blob/be3ac266a50f4884bc2d5cbf471cd43aab1a2cef/mmlearn/tasks/ijepa_pretraining.py#L106)):*
> ```Python
> 
>         if checkpoint_path != "":
>             self.encoder, self.predictor, self.target_encoder, _, _, _ = (
> ```

Also, as we discussed, I think each module should handle loading the pretrained checkpoints individually. An example use case for why is being able to use iJEPA-pretrained encoder (the original one) for contrastive pretraining.
___
*[`mmlearn/tasks/ijepa_pretraining.py` line 171 at r1](https://reviewable.io/reviews/VectorInstitute/mmlearn/25#-O8wctiB1lE_PFpds8C-:-O8wctiB1lE_PFpds8C0:b-ob7vtb) ([raw file](https://github.com/VectorInstitute/mmlearn/blob/be3ac266a50f4884bc2d5cbf471cd43aab1a2cef/mmlearn/tasks/ijepa_pretraining.py#L171)):*
> ```Python
>         return encoder, predictor, target_encoder, opt, scaler, epoch
> 
>     def forward(
> ```

This forward pass is currently not being used by this module. Everything is in `training_step` right now. Remember that the LightningModule is also an nn.Module.
___
*[`mmlearn/tasks/ijepa_pretraining.py` line 177 at r1](https://reviewable.io/reviews/VectorInstitute/mmlearn/25#-O8wbaGE0V_o274ZnUOI:-O8wbaGE0V_o274ZnUOJ:bqboadp) ([raw file](https://github.com/VectorInstitute/mmlearn/blob/be3ac266a50f4884bc2d5cbf471cd43aab1a2cef/mmlearn/tasks/ijepa_pretraining.py#L177)):*
> ```Python
>     ) -> torch.Tensor:
>         """Forward pass through the encoder."""
>         return self.encode(x, masks)
> ```

The `encode` method doesn't exist. Please make sure to try running the code to ensure that things are working properly.
Also, please take a look at the other encoders in the library - the current convention is to pass the entire batch dictionary to the encoder.
___
*[`mmlearn/tasks/ijepa_pretraining.py` line 181 at r1](https://reviewable.io/reviews/VectorInstitute/mmlearn/25#-O8wbFYsCfVj3hGOdOGG:-O8wbFYsCfVj3hGOdOGH:b-dupd2s) ([raw file](https://github.com/VectorInstitute/mmlearn/blob/be3ac266a50f4884bc2d5cbf471cd43aab1a2cef/mmlearn/tasks/ijepa_pretraining.py#L181)):*
> ```Python
>     def training_step(self, batch: Dict[str, Any], batch_idx: int) -> torch.Tensor:
>         """Perform a single training step."""
>         images = batch[Modalities.RGB]
> ```

The format for this has changed. See [this PR](https://github.com/VectorInstitute/mmlearn/pull/26).



_Suggestion:_
```Python
images = batch[Modalities.RGB.name]
```
___
*[`mmlearn/tasks/ijepa_pretraining.py` line 191 at r1](https://reviewable.io/reviews/VectorInstitute/mmlearn/25#-O8wc7wkCvWx6i8Aiin1:-O8wc7wkCvWx6i8Aiin2:b-ocrfdd) ([raw file](https://github.com/VectorInstitute/mmlearn/blob/be3ac266a50f4884bc2d5cbf471cd43aab1a2cef/mmlearn/tasks/ijepa_pretraining.py#L191)):*
> ```Python
> 
>         # Move images and masks to device
>         images = images.to(self.device)
> ```

You don't need to do this for anything inside the batch dictionary. Lightning will handle that.
___
*[`mmlearn/tasks/ijepa_pretraining.py` line 216 at r1](https://reviewable.io/reviews/VectorInstitute/mmlearn/25#-O8wcSAP7VBaahGXpli5:-O8wcSAP7VBaahGXpli6:b-axpwjf) ([raw file](https://github.com/VectorInstitute/mmlearn/blob/be3ac266a50f4884bc2d5cbf471cd43aab1a2cef/mmlearn/tasks/ijepa_pretraining.py#L216)):*
> ```Python
>             "train/loss",
>             loss,
>             on_step=True,
> ```

Why log both on_step and on_epoch? During training, the default for lightning is to log on_step but not on_epoch (for validation and testing, the default is to log on_epoch but not on_step)
___
*[`mmlearn/tasks/ijepa_pretraining.py` line 227 at r1](https://reviewable.io/reviews/VectorInstitute/mmlearn/25#-O8wdGpu1JQUKe2XGTGh:-O8wdGpv2jr9WwxtZsSA:b30t9x0) ([raw file](https://github.com/VectorInstitute/mmlearn/blob/be3ac266a50f4884bc2d5cbf471cd43aab1a2cef/mmlearn/tasks/ijepa_pretraining.py#L227)):*
> ```Python
>         return loss
> 
>     def _update_target_encoder(self) -> None:
> ```

Please look into reusing the existing EMA module.
___
*[`mmlearn/tasks/ijepa_pretraining.py` line 247 at r1](https://reviewable.io/reviews/VectorInstitute/mmlearn/25#-O8wdMQ77HhYudRGLl6P:-O8wdMQ77HhYudRGLl6Q:bp998or) ([raw file](https://github.com/VectorInstitute/mmlearn/blob/be3ac266a50f4884bc2d5cbf471cd43aab1a2cef/mmlearn/tasks/ijepa_pretraining.py#L247)):*
> ```Python
>                 "params": (p for n, p in self.encoder.named_parameters()
>                            if ("bias" not in n) and (len(p.shape) != 1)),
>                 "weight_decay": self.optimizer_cfg.get("weight_decay", 0.0)
> ```

Like I mentioned earlier, you will be getting an instantiated Optimizer object, not the config. Please take a look at the contrastive pretraining task for how to get the weight decay value from the instantiated Optimizer object.
___
*[`mmlearn/tasks/ijepa_pretraining.py` line 309 at r1](https://reviewable.io/reviews/VectorInstitute/mmlearn/25#-O8wdlb18ouNLGAh0fsN:-O8wdlb2DHYpA7Aq-hOq:b-juppwq) ([raw file](https://github.com/VectorInstitute/mmlearn/blob/be3ac266a50f4884bc2d5cbf471cd43aab1a2cef/mmlearn/tasks/ijepa_pretraining.py#L309)):*
> ```Python
>         return self._shared_eval_step(batch, batch_idx, "test")
> 
>     def _shared_eval_step(
> ```

Notice that most of the code in this method is repeated in `training_step`. You can define it once and call it multiple times for training and eval.

**Review by fcogidi — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/VectorInstitute/mmlearn/25)* status: 0 of 2 files reviewed, 4 unresolved discussions (waiting on @vahid0001)

___
*[`mmlearn/modules/encoders/vision.py` line 313 at r2](https://reviewable.io/reviews/VectorInstitute/mmlearn/25#-O9pnHatDHwCSWdTCbx2:-O9pnHatDHwCSWdTCbx3:b-6e3w6z) ([raw file](https://github.com/VectorInstitute/mmlearn/blob/41a21f69f17b949f513905f01f215947317b1e70/mmlearn/modules/encoders/vision.py#L313)):*
> ```Python
>                 nn.init.constant_(m.bias, 0)
> 
>     def load_checkpoint(
> ```

How's this intended to be used? 

It looks like both the encoder and predictor needs to be instantiated first and then passed to this method. When I had first made the suggestion to move the checkpoint loading functionality to the encoder, I imagined the `VisionTransformer` having its own checkpoint loading logic (just for the encoder) and the predictor having its own (loading the target encoder may not be necessary, I think, if one knows the exact ema value where it stopped at).
___
*[`mmlearn/tasks/ijepa_pretraining.py` line 79 at r2](https://reviewable.io/reviews/VectorInstitute/mmlearn/25#-O9pjDPFBcvgTIHJnHrU:-O9pjDPFBcvgTIHJnHrV:b-86vqeu) ([raw file](https://github.com/VectorInstitute/mmlearn/blob/41a21f69f17b949f513905f01f215947317b1e70/mmlearn/tasks/ijepa_pretraining.py#L79)):*
> ```Python
>         self.predictor = predictor
> 
>         self.ema = ExponentialMovingAverage(encoder, ema_decay, ema_decay_end, 1000)
> ```

Should the `ema_anneal_end_step`, currently fixed at `1000` be a user-defined value?

Also, `device_id` might be important to set, especially in distributed setting. It can be set to `self.device`.
___
*[`mmlearn/tasks/ijepa_pretraining.py` line 138 at r2](https://reviewable.io/reviews/VectorInstitute/mmlearn/25#-O9pmVTOCh3Kxy_7n6_W:-O9pmVTOCh3Kxy_7n6_X:b1c3bcl) ([raw file](https://github.com/VectorInstitute/mmlearn/blob/41a21f69f17b949f513905f01f215947317b1e70/mmlearn/tasks/ijepa_pretraining.py#L138)):*
> ```Python
>         )
> 
>         if is_training:
> ```

Is the `is_training` flag necessary? Why not, `if step_type == "train"`?
___
*[`mmlearn/tasks/ijepa_pretraining.py` line 230 at r2](https://reviewable.io/reviews/VectorInstitute/mmlearn/25#-O9pmuFaB_Jpzb8UTyjE:-O9pmuFaB_Jpzb8UTyjF:bvt6nuc) ([raw file](https://github.com/VectorInstitute/mmlearn/blob/41a21f69f17b949f513905f01f215947317b1e70/mmlearn/tasks/ijepa_pretraining.py#L230)):*
> ```Python
>     def on_validation_epoch_start(self) -> None:
>         """Prepare for the validation epoch."""
>         self._on_eval_epoch_start("val")
> ```

The `self._on_eval_epoch_{start/end}` methods are not defined.

**Review by fcogidi — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/VectorInstitute/mmlearn/25)* status: 0 of 2 files reviewed, 4 unresolved discussions (waiting on @vahid0001)

___
*[`mmlearn/modules/encoders/vision.py` line 313 at r2](https://reviewable.io/reviews/VectorInstitute/mmlearn/25#-O9pnHatDHwCSWdTCbx2:-O9ppF1J1Jc4dtSxeiPT:b1tpklt) ([raw file](https://github.com/VectorInstitute/mmlearn/blob/41a21f69f17b949f513905f01f215947317b1e70/mmlearn/modules/encoders/vision.py#L313)):*
<details><summary><i>Previously, fcogidi (Franklin) wrote…</i></summary><blockquote>

How's this intended to be used? 

It looks like both the encoder and predictor needs to be instantiated first and then passed to this method. When I had first made the suggestion to move the checkpoint loading functionality to the encoder, I imagined the `VisionTransformer` having its own checkpoint loading logic (just for the encoder) and the predictor having its own (loading the target encoder may not be necessary, I think, if one knows the exact ema value where it stopped at).
</blockquote></details>

Speaking of, please check [this](https://github.com/VectorInstitute/Multimodal/blob/medmultimodal/multimode/tasks/data2vec.py#L559) out.

**Review by fcogidi — APPROVED** [review]
